File: CS320/HW1/4.hs
-- Teng Xu (xt@bu.edu)
-- 4.1
recurfib :: [Int] -> Int -> Int
recurfib lst k
    | k > length lst = recurfib lst (length lst)
    | k == 0 = 0
    | otherwise = (head lst) + recurfib (tail lst) (k-1)

--4.2
recurfib2help :: Int -> Int -> [Int] -> Int
recurfib2help n k lst
    |n == 0 = 1
    |n == 1 = 1
    |n == 2 = 1
    |n == length lst = recurfib lst k
    |otherwise = recurfib2help n k ((recurfib lst k) : lst)

recurfib2 :: Int -> Int -> Int
recurfib2 n k = recurfib2help n k [1,1,0]

File: CS320/HW1/4.py
# 4.1
def recurfib (n, k):
    if k > len(n):
        return recurfib(n, len(n))
    elif k == 0:
        return 0
    else:
        return n[0] + recurfib(n[1:],k-1)

# 4.2
def recurfib2(n,k):
    return recurfib2help(n,k,[1,1,0])

def recurfib2help (n, k, lst):
    if(n == 0):
        return 1
    elif (n==1):
        return 1
    elif (n==2):
        return 1
    elif (n == len(lst)):
        return recurfib(lst,k)
    else:
        return recurfib2help(n,k,[recurfib(lst,k)]+lst)


